src/h3lib/lib/area.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2025 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @file area.c
+ * @brief   Algorithms for computing areas of regions on a sphere (GeoLoop,
+ * cells, polygons, multipolygons, etc.)
+ */
+
+#include "area.h"
+
+#include <math.h>
+
+#include "adder.h"
+#include "constants.h"
+#include "h3Assert.h"
+#include "h3api.h"
+
+// Cagnoli contribution for edge arc x to y, following d3-geo’s
+// area implementation:
+// https://github.com/d3/d3-geo/blob/8c53a90ae70c94bace73ecb02f2c792c649c86ba/src/area.js#L51-L70
+static inline double cagnoli(LatLng x, LatLng y) {
+    x.lat = x.lat / 2.0 + M_PI / 4.0;
+    y.lat = y.lat / 2.0 + M_PI / 4.0;
+
+    double sa = sin(x.lat) * sin(y.lat);
+    double ca = cos(x.lat) * cos(y.lat);
+
+    double d = y.lng - x.lng;
+    double sd = sin(d);
+    double cd = cos(d);
+
+    return -2.0 * atan2(sa * sd, sa * cd + ca);
+}
+
+/**
+ * Area in radians^2 enclosed by vertices in GeoLoop.
+ *
+ * The GeoLoop should represent a simple curve with no self-intersections.
+ * Vertices should be ordered according to the "right hand rule".
+ * That is, if you are looking from outer space at a spherical polygon loop
+ * on the surface of the earth whose interior is contained within a hemisphere,
+ * then the vertices should be ordered counter-clockwise. The interior of the
+ * loop is to the left of a person walking along the boundary of the polygon
+ * in the counter-clockwise direction.
+ *
+ * Note that GeoLoops do not need to repeat the first vertex at the end of the
+ * array to close the loop; this is done automatically.
+ *
+ * The edge arcs between adjacent vertices are assumed to be the shortest
+ * geodesic path between them; that is, all arcs are interpreted to be less
+ * than 180 degrees or pi radians.
+ * Avoid arcs that are exactly pi (i.e, two antipodal vertices).
```

Comment:
Might want to add a comment on consequences of this, and maybe throw an error if such an arc is encountered? Though presumably you could only tell if they were using the same const for Pi.

Reply:
There isn't a numerically stable test for *exactly* pi radians apart, so I'm not sure what we can do other than to let users know that two points will always be interpreted as describing the shortest arc between them. And to note that there are discontinuities at around arcs with pi radians, so they should be avoided. 

I suppose one thing we could do would be to pick some tolerance and throw an error if an arc is within `pi - tol`. Let me create an issue for that.